encoding/encoding_test.go

```diff
@@ -53,3 +82,303 @@ func TestDuplicateCompressorRegister(t *testing.T) {
+// that encoding and decoding happen once per RPC.
+func (s) TestForceServerCodec(t *testing.T) {
+	// Create an server with the counting proto codec.
+	codec := &countingProtoCodec{name: t.Name()}
+	backend := stubserver.StartTestService(t, nil, grpc.ForceServerCodec(codec))
+	defer backend.Stop()
+
+	// Create a channel to the above server.
+	cc, err := grpc.Dial(backend.Address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("Failed to dial test backend at %q: %v", backend.Address, err)
+	}
+	defer cc.Close()
+
+	// Make an RPC and expect it to fail. Since we do not specify any codec
+	// here, the proto codec will get automatically used.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	client := testgrpc.NewTestServiceClient(cc)
+	if _, err := client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+		t.Fatalf("ss.Client.EmptyCall(_, _) = _, %v; want _, nil", err)
+	}
+
+	unmarshalCount := atomic.LoadInt32(&codec.unmarshalCount)
+	const wantUnmarshalCount = 1
+	if unmarshalCount != wantUnmarshalCount {
+		t.Fatalf("Unmarshal Count = %d; want %d", unmarshalCount, wantUnmarshalCount)
+	}
+	marshalCount := atomic.LoadInt32(&codec.marshalCount)
+	const wantMarshalCount = 1
+	if marshalCount != wantMarshalCount {
+		t.Fatalf("MarshalCount = %d; want %d", marshalCount, wantMarshalCount)
+	}
+}
+
+// renameProtoCodec wraps the proto codec and allows customizing the Name().
+type renameProtoCodec struct {
+	encoding.Codec
+	name string
+}
+
+func (r *renameProtoCodec) Name() string { return r.name }
+
+// TestForceCodecName confirms that the ForceCodec call option sets the subtype
+// in the content-type header according to the Name() of the codec provided.
+func (s) TestForceCodecName(t *testing.T) {
+	wantContentTypeCh := make(chan []string, 1)
+	defer close(wantContentTypeCh)
+
+	// Create a test service backend that pushes the received content-type on a
+	// channel for the test to inspect.
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			md, ok := metadata.FromIncomingContext(ctx)
+			if !ok {
+				return nil, status.Errorf(codes.Internal, "no metadata in context")
+			}
+			if got, want := md["content-type"], <-wantContentTypeCh; !cmp.Equal(got, want) {
+				return nil, status.Errorf(codes.Internal, "got content-type=%q; want [%q]", got, want)
+			}
+			return &testpb.Empty{}, nil
+		},
+	}
+	// Since we don't specify a codec as a server option, it will end up
+	// automatically using the proto codec.
+	if err := ss.Start(nil); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Force the use of the custom codec on the client with the ForceCodec call
+	// option. Confirm the name is converted to lowercase before transmitting.
```

Comment:
nit: i feel like this could go to the doc comments since this seems like a pretty important assertion made in this test.

Reply:
I moved it to the test level comment. The case insensitivity is part of the public godoc: https://github.com/grpc/grpc-go/blob/master/encoding/encoding.go#L99-L112